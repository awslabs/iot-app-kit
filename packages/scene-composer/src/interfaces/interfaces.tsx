/**
 * This file contains public interfaces of the scene content the SceneComposer exposes.
 * It is different from the serialization format from the GLTF Toolkit package.
 * These interfaces are used to interact with the SceneComposer instance at
 * runtime, thus the data structure is slightly different from the GLTF Toolkit.
 */

import { ReactElement } from 'react';

import { ISceneNodeInternal } from '../store/Store';

import { ISceneComponent, KnownComponentType } from './components';
import { IRuleBasedMap, ITransform, ITransformConstraint, Vector3 } from './dataTypes';

/************************************************
 * Scene document, entity, component models
 ************************************************/

export interface ISceneDocumentSnapshot {
  /**
   * Serialize the current document snapshot into a document string with the
   * given format version.
   *
   * Note that this interface is not optimized for very large scene, which
   * we may introduce new APIs later to use stream API to handle the output.
   *
   * TODO: do we want to allow updating the document version?
   * [Option 1]: Externalize the document version instea of keeping the doucment version in the document itself.
   * [Option 2]: Allow passing in a parameter to set the document version.
   * [Option 3]: Automatically increment the document version by 1 based on the original input document version.
   *
   * We'll go with option 1 for now for simplicity.
   *
   * TODO: what should we return when the scene is empty?
   * [Option 1]: Return a falsy value, such as empty string.
   * [Option 2]: Return a valid scene with no entity.
   *
   * We'll go with option 2 for now.
   *
   * @param specVersion - version of the document format
   */
  serialize(specVersion: string, ...stringifyArgs): string;
}

export interface ISceneDocument {
  nodeMap: Record<string, ISceneNode>;
  ruleMap: Record<string, IRuleBasedMap>;
  rootNodeRefs: string[];
  unit?: string;
  version: string;
  properties?: Partial<Record<KnownSceneProperty, any>>;
  specVersion?: string;
}

export interface ISceneNode {
  // ref is optional when used in APIs to create nodes
  ref?: string;
  name?: string;
  transform?: ITransform;
  transformConstraint?: ITransformConstraint;
  components?: ISceneComponent[];
  parentRef?: string;
  childRefs?: string[];
  properties?: Partial<Record<'alwaysVisible' | 'matterportId' | 'hiddenWhileImmersive', boolean | string>>;
}

export enum KnownSceneProperty {
  BaseUrl = 'baseUrl',
  EnvironmentPreset = 'environmentPreset',
  DataBindingConfig = 'dataBindingConfig',
  ComponentSettings = 'componentSettings',
  MatterportModelId = 'matterportModelId',
  LayerIds = 'layerIds',
  SceneRootEntityId = 'sceneRootEntityId',
  TagCustomColors = 'tagCustomColors',
  FogSettings = 'fogSettings',
}

/************************************************
 * Scene Composer Control APIs
 ************************************************/

export type CameraControlMode = 'transition' | 'teleport';
export type CameraControlsType = 'orbit' | 'pan' | 'immersive' | 'pointerLock';
export type TransformControlMode = 'translate' | 'rotate' | 'scale';
export enum CameraViewAxisValues {
  Front = 'front',
  Back = 'back',
  Left = 'left',
  Right = 'right',
  Top = 'top',
  Bottom = 'bottom',
}

// an object ref or a precise camera setting
export type FixedCameraTarget = { position: Vector3; target: Vector3 };
export type NamedCameraTarget = string;
export type CameraTarget = NamedCameraTarget | FixedCameraTarget;

export type URIModifier = (uri: string) => string;

export type CreateErrorViewCallback = (error: Error) => ReactElement;

/************************************************
 * Scene Namespace
 ************************************************/

const IotTwinMakerNamespace = 'iottwinmaker.common';
export const IotTwinMakerNamespaceSeparator = ':';
export const IotTwinMakerIconNamespace = `${IotTwinMakerNamespace}.icon`;
export const IotTwinMakerColorNamespace = `${IotTwinMakerNamespace}.color`;
export const IotTwinMakerNumberNamespace = `${IotTwinMakerNamespace}.number`;
export const IotTwinMakerOpacityNamespace = `${IotTwinMakerNamespace}.opacity`;
/************************************************
 * Scene Resource
 ************************************************/

export enum SceneResourceType {
  Icon = 'Icon',
  Color = 'Color',
  Number = 'Number',
  Opacity = 'Opacity',
}

export interface SceneResourceInfo {
  type: SceneResourceType;
  value: string;
}

export interface AddingWidgetInfo {
  type?: KnownComponentType;
  node: ISceneNodeInternal;
}

export interface MeshStyle {
  color?: string | number;
  opacity?: number;
  transparent?: boolean;
}

export interface StyleTarget {
  dataBindingContext: unknown;
  style: MeshStyle;
}

export interface IFogSettings {
  colorHex: number;
  near: number;
  far: number;
}
