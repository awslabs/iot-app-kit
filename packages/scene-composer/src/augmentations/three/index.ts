export * from './Anchor';
export * from './Viewpoint';
