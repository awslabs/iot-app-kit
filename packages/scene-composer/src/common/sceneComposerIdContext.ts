import * as React from 'react';

export const sceneComposerIdContext = React.createContext('default');

export const useSceneComposerId = () => {
  return React.useContext(sceneComposerIdContext);
};
