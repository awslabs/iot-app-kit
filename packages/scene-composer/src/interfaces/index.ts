export * from './assets';
export * from './interfaces';
export * from './components';
export * from './dataTypes';
export * from './dataBinding';
export * from './metricRecorder';
export * from './feature';
export * from './sceneViewer';
export * from './componentSettings';
