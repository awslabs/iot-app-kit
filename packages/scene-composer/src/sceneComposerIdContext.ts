import * as React from 'react';

export const sceneComposerIdContext = React.createContext('default');
