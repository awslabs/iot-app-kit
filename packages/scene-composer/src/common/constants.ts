import * as THREE from 'three';

import { Component, LightType } from '../models/SceneModels';
import { InfoIconSvgString, WarningIconSvgString, ErrorIconSvgString, VideoIconSvgString } from '../assets';
import {
  IValueDataBindingProviderState,
  DefaultAnchorStatus,
  DistanceUnit,
  Vector3,
  ITagSettings,
  IOverlaySettings,
} from '../interfaces';
import { CameraControlImpl } from '../store/internalInterfaces';

/******************************************************************************
 * Document Constants
 ******************************************************************************/

// TODO: [4553] remove in preview build
export const LEGACY_VERSION = '1';

export const CURRENT_MAJOR_VERSION = 1;
export const CURRENT_MINOR_VERSION = 0;
export const CURRENT_VERSION = `${CURRENT_MAJOR_VERSION}.${CURRENT_MINOR_VERSION}`;

// TODO: [4812] replace the url with the actual one
export const AWS_USER_GUIDE_DOC_URL = 'https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html';

/******************************************************************************
 * DataBinding Constants
 ******************************************************************************/
export const DataBindingLabelKeys = {
  entityId: 'entityId',
  componentName: 'componentName',
  propertyName: 'propertyName',
  entityPath: 'entityPath',
  componentTypeId: 'componentTypeId',
};

export const EMPTY_VALUE_DATA_BINDING_PROVIDER_STATE: IValueDataBindingProviderState = {
  definitions: [],
  selectedOptions: [],
};

// For V1, the customer can only use 2 predefined data binding template in scene editor
export const DEFAULT_DATA_BINDING_TEMPLATE_ENTITY_ID = 'sel_entity';
export const DEFAULT_DATA_BINDING_TEMPLATE_COMPONENT_NAME = 'sel_comp';

/******************************************************************************
 * Scene Constants
 ******************************************************************************/
export const DEFAULT_DISTANCE_UNIT: DistanceUnit = 'meters';

export const ROOT_OBJECT_3D_NAME = 'ROOT';

export const DEFAULT_COLOR_MAPS = [
  new THREE.Color('red'),
  new THREE.Color('blue'),
  new THREE.Color('green'),
  new THREE.Color('crimson'),
  new THREE.Color('cyan'),
];

export const DEFAULT_CAMERA_SETTINGS = {
  fov: 53.13,
  near: 0.1,
  far: 1000.0,
  zoom: 1,
};

const DEFAULT_DIRECTIONAL_LIGHT_SETTINGS: Component.IDirectionalLightSettings = {
  color: 0xffffff,
  intensity: 1,
  castShadow: true,
};

const DEFAULT_POINT_LIGHT_SETTINGS: Component.IPointLightSettings = {
  color: 0xffffff,
  intensity: 1,
  decay: 2.0,
  distance: 0,
  castShadow: true,
};

const DEFAULT_AMBIENT_LIGHT_SETTINGS: Component.IAmbientLightSettings = {
  color: 0xffffff,
  intensity: 1,
};

const DEFAULT_HEMISPHERE_LIGHT_SETTINGS: Component.IHemisphereLightSettings = {
  color: 0xffffff,
  groundColor: 0x333333,
  intensity: 1,
};

export const DEFAULT_LIGHT_SETTINGS_MAP = {
  [LightType.Directional]: DEFAULT_DIRECTIONAL_LIGHT_SETTINGS,
  [LightType.Point]: DEFAULT_POINT_LIGHT_SETTINGS,
  [LightType.Hemisphere]: DEFAULT_HEMISPHERE_LIGHT_SETTINGS,
  [LightType.Ambient]: DEFAULT_AMBIENT_LIGHT_SETTINGS,
};

export const SCENE_ICONS: Record<DefaultAnchorStatus, string> = {
  [DefaultAnchorStatus.Info]: InfoIconSvgString,
  [DefaultAnchorStatus.Warning]: WarningIconSvgString,
  [DefaultAnchorStatus.Error]: ErrorIconSvgString,
  [DefaultAnchorStatus.Video]: VideoIconSvgString,
};

export enum RenderOrder {
  DrawEarly = -10,
  Default = 0,
  DrawLate = 10,
}

export enum Layers {
  RaycastAndRender = 0,
  RenderOnly = 1,
}

export const SCENE_BODY_CLASS = 'twinmaker_scene_container';

export const DRACO_PATH = 'https://www.gstatic.com/draco/versioned/decoders/1.4.1/';

export const DEFAULT_TAG_GLOBAL_SETTINGS: ITagSettings = {
  autoRescale: false,
  scale: 1,
};

export const DEFAULT_OVERLAY_GLOBAL_SETTINGS: IOverlaySettings = {
  overlayPanelVisible: false,
};

/******************************************************************************
 * Camera Constants
 ******************************************************************************/
export const DEFAULT_CAMERA_CONTROLS_OPTIONS: Pick<CameraControlImpl, 'dampingFactor' | 'maxDistance' | 'minDistance'> =
  {
    dampingFactor: 0.2,
    maxDistance: Infinity,
    minDistance: 0,
  };
export const DEFAULT_CAMERA_POSITION: Vector3 = [5, 5, 5];
export const DEFAULT_CAMERA_OPTIONS: Pick<THREE.PerspectiveCamera, 'far' | 'fov' | 'near'> = {
  far: 1000,
  fov: 53.13,
  near: 0.1,
};
export const DEFAULT_CAMERA_TARGET: Vector3 = [0, 0, 0];
export const DEFAULT_TWEEN_DURATION = 500;
export const SCENE_CAPABILITY_MATTERPORT = 'MATTERPORT';
export const SECRET_MANAGER_MATTERPORT_TAG = 'AWSIoTTwinMaker_Matterport';
export const MATTERPORT_SECRET_ARN = 'MATTERPORT_SECRET_ARN';
export const MATTERPORT_ACCESS_TOKEN = 'MATTERPORT_ACCESS_TOKEN';
export const MATTERPORT_APPLICATION_KEY = 'MATTERPORT_APPLICATION_KEY';
export const OPTIONS_PLACEHOLDER_VALUE = 'n/a';
