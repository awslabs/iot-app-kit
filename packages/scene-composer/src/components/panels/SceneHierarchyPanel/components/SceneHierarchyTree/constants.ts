import { KnownComponentType } from '../../../../../interfaces';
import draggable from '../../../../../enhancers/draggable';
import droppable from '../../../../../enhancers/droppable';
import Tree, { TreeItem } from '../../../../Tree';

export const AcceptableDropTypes = [
  ...Object.keys(KnownComponentType).filter((t) => t !== KnownComponentType.SubModelRef),
  'Space',
  'default',
];

export const EnhancedTreeItem = droppable(draggable(TreeItem));
export const EnhancedTree = droppable(Tree);
