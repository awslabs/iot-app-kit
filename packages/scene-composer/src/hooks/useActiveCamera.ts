import { useCallback } from 'react';
import * as THREE from 'three';

import { useSceneComposerId } from '../common/sceneComposerIdContext';
import { useEditorState } from '../store';
import { CameraControlMode, CameraSettings } from '../interfaces';
import { DEFAULT_CAMERA_OPTIONS, DEFAULT_CAMERA_POSITION, DEFAULT_CAMERA_TARGET } from '../common/constants';
import { CameraType, Transform } from '../models/SceneModels';

const useActiveCamera = () => {
  const sceneComposerId = useSceneComposerId();
  const {
    activeCameraSettings,
    setActiveCameraSettings,
    setCameraTarget,
    isViewing,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  } = useEditorState(sceneComposerId);

  const defaultCameraSettings: CameraSettings = {
    cameraType: CameraType.Perspective,
    fov: DEFAULT_CAMERA_OPTIONS.fov,
    far: DEFAULT_CAMERA_OPTIONS.far,
    near: DEFAULT_CAMERA_OPTIONS.near,
    zoom: 1,
    transform: {
      position: DEFAULT_CAMERA_POSITION,
    },
  };

  const calculateTargetOrGetDefault = (transform: Transform) => {
    if (transform.rotation) {
      const obj3D = new THREE.Object3D();
      obj3D.position.set(...transform.position);
      obj3D.rotation.set(...transform.rotation);
      obj3D.updateMatrix();
      obj3D.updateMatrixWorld(true);

      // Place an object to look at in front of the parent and get it's world position
      const targetObj3D = new THREE.Object3D();
      obj3D.add(targetObj3D);
      targetObj3D.position.set(0, 0, -10);
      targetObj3D.updateMatrix();
      targetObj3D.updateMatrixWorld(true);
      const target = targetObj3D.getWorldPosition(new THREE.Vector3());
      return target.toArray();
    }

    return DEFAULT_CAMERA_TARGET;
  };

  const setActiveCamera = useCallback(
    (cameraSettings?: CameraSettings, mode: CameraControlMode = 'transition', forceSetTarget = false) => {
      const settings = cameraSettings ?? defaultCameraSettings;
      setActiveCameraSettings(settings);

      const cameraTarget = {
        position: settings.transform.position,
        target: calculateTargetOrGetDefault(settings.transform),
      };

      if (forceSetTarget || isViewing()) {
        setCameraTarget(cameraTarget, mode);
      }
    },
    [setActiveCameraSettings, setCameraTarget],
  );

  return {
    activeCameraSettings,
    setActiveCameraSettings: setActiveCamera,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  };
};

export default useActiveCamera;
