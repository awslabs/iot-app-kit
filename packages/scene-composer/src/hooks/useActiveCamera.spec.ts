import { MathUtils } from 'three';

import { useEditorState } from '../store';
import { useSceneComposerId } from '../common/sceneComposerIdContext';
import { CameraSettings } from '../interfaces';
import { DEFAULT_CAMERA_OPTIONS, DEFAULT_CAMERA_POSITION, DEFAULT_CAMERA_TARGET } from '../common/constants';

import useActiveCamera from './useActiveCamera';

jest.mock('../store', () => ({
  ...jest.requireActual('../store'),
  useEditorState: jest.fn(),
}));

jest.mock('../common/sceneComposerIdContext', () => ({
  ...jest.requireActual('../common/sceneComposerIdContext'),
  useSceneComposerId: jest.fn(),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useCallback: jest.fn((cb) => cb), // unbox callbacks.
}));

describe('useSelectedObject', () => {
  const defaultCameraSettings: CameraSettings = {
    cameraType: 'Perspective',
    fov: DEFAULT_CAMERA_OPTIONS.fov,
    far: DEFAULT_CAMERA_OPTIONS.far,
    near: DEFAULT_CAMERA_OPTIONS.near,
    zoom: 1,
    transform: {
      position: DEFAULT_CAMERA_POSITION,
    },
  };
  const sceneComposerId = 'default';
  const setActiveCameraSettings = jest.fn();
  const isViewing = jest.fn();
  const setCameraTarget = jest.fn();

  beforeEach(() => {
    (useSceneComposerId as jest.Mock).mockImplementation(() => sceneComposerId);
    (useEditorState as jest.Mock).mockImplementation(() => ({
      setActiveCameraSettings,
      activeCameraSettings: defaultCameraSettings,
      isViewing,
      setCameraTarget,
    }));
    jest.clearAllMocks();
  });

  it('should return the object from the store', () => {
    // Act
    const result = useActiveCamera();

    expect(result).toMatchSnapshot();
  });

  it('should call dependencies when setActiveCameraSettings is called and in viewing mode', () => {
    const differentCameraSettings: CameraSettings = {
      cameraType: 'Perspective',
      fov: 100,
      far: 20,
      near: 0.1,
      zoom: 1,
      transform: {
        position: [1, 1, 1],
      },
    };
    isViewing.mockReturnValue(true);

    const activeCamera = useActiveCamera();
    activeCamera.setActiveCameraSettings(differentCameraSettings, 'teleport');

    expect(setActiveCameraSettings).toHaveBeenCalledWith(differentCameraSettings);
    expect(setCameraTarget).toHaveBeenCalledWith({ position: [1, 1, 1], target: DEFAULT_CAMERA_TARGET }, 'teleport');
  });

  it('should call dependencies with calculated target is called and in viewing mode with a rotation transform', () => {
    const differentCameraSettings: CameraSettings = {
      cameraType: 'Perspective',
      fov: 1000,
      far: 50,
      near: 0.2,
      zoom: 1,
      transform: {
        position: [10, 0, 0],
        rotation: [0, MathUtils.degToRad(90), 0],
      },
    };
    const expectedCalculatedTarget = [0, 0, -2.220446049250313e-15];
    isViewing.mockReturnValue(true);

    const activeCamera = useActiveCamera();
    activeCamera.setActiveCameraSettings(differentCameraSettings);

    expect(setActiveCameraSettings).toHaveBeenCalledWith(differentCameraSettings);
    expect(setCameraTarget).toHaveBeenCalledWith(
      { position: [10, 0, 0], target: expectedCalculatedTarget },
      'transition',
    );
  });

  it('should call setActiveCameraSettings but not setCameraTarget when setActiveCameraSettings is called and in edit mode', () => {
    const differentCameraSettings: CameraSettings = {
      cameraType: 'Perspective',
      fov: 100,
      far: 20,
      near: 0.1,
      zoom: 1,
      transform: {
        position: [1, 1, 1],
      },
    };
    isViewing.mockReturnValue(false);

    const activeCamera = useActiveCamera();
    activeCamera.setActiveCameraSettings(differentCameraSettings, 'teleport');

    expect(setActiveCameraSettings).toHaveBeenCalledWith(differentCameraSettings);
    expect(setCameraTarget).not.toBeCalled();
  });

  it('should call dependencies when setActiveCameraSettings is called and in edit mode when forced', () => {
    const differentCameraSettings: CameraSettings = {
      cameraType: 'Perspective',
      fov: 100,
      far: 20,
      near: 0.1,
      zoom: 1,
      transform: {
        position: [1, 1, 1],
      },
    };
    isViewing.mockReturnValue(false);

    const activeCamera = useActiveCamera();
    activeCamera.setActiveCameraSettings(differentCameraSettings, 'teleport', true);

    expect(setActiveCameraSettings).toHaveBeenCalledWith(differentCameraSettings);
    expect(setCameraTarget).toHaveBeenCalledWith({ position: [1, 1, 1], target: DEFAULT_CAMERA_TARGET }, 'teleport');
  });
});
