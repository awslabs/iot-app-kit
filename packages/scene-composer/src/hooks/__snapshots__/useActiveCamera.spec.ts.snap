// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSelectedObject should return the object from the store 1`] = `
{
  "activeCameraName": undefined,
  "activeCameraSettings": {
    "cameraType": "Perspective",
    "far": 1000,
    "fov": 53.13,
    "near": 0.1,
    "transform": {
      "position": [
        5,
        5,
        5,
      ],
    },
    "zoom": 1,
  },
  "mainCameraObject": undefined,
  "setActiveCameraName": undefined,
  "setActiveCameraSettings": [Function],
}
`;
